File: examples/android/MSDKSampleAppJava/app/src/main/java/msdk/java/sample/history/HistoryHandler.java
package msdk.java.sample.history;

import java.util.concurrent.Executors;

import msdk.java.sample.SingleLiveData;
import msdk.java.sample.db.Database;
import msdk.java.sample.db.entity.Action;
import msdk.java.sample.homepage.Results;

import static msdk.java.sample.db.ActionStatus.HISTORIZED;
import static msdk.java.sample.homepage.Results.ACTION_FAILURE;
import static msdk.java.sample.homepage.Results.ACTION_SUCCESS;
import static msdk.java.sample.homepage.Results.FAILURE;
import static msdk.java.sample.homepage.Results.REJECT;
import static msdk.java.sample.homepage.Results.SUCCESS;

public class HistoryHandler {

    public static void addHistoryAction(
            Database db,
            String name,
            String description,
            String icon,
            SingleLiveData<Results> liveData
    ) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                Action action = new Action();
                action.invite = null;
                action.name = name;
                action.description = description;
                action.icon = icon;
                action.status = HISTORIZED.toString();
                db.actionDao().insertAll(action);
                liveData.postValue(ACTION_SUCCESS);
            } catch (Exception e) {
                e.printStackTrace();
                liveData.postValue(ACTION_FAILURE);
            }
        });
    }

    public static void addToHistory(int actionId, String description, Database db, SingleLiveData<Results> liveData) {
        try {
            Action action = db.actionDao().getActionsById(actionId);
            action.status = HISTORIZED.toString();
            action.description = description;
            db.actionDao().update(action);
            liveData.postValue(SUCCESS);
        } catch (Exception e) {
            e.printStackTrace();
            liveData.postValue(FAILURE);
        }
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/history/HistoryHandler.kt
package msdk.kotlin.sample.history

import msdk.kotlin.sample.SingleLiveData
import msdk.kotlin.sample.db.ActionStatus
import msdk.kotlin.sample.db.Database
import msdk.kotlin.sample.db.entity.Action
import msdk.kotlin.sample.homepage.Results

object HistoryHandler {
    suspend fun addHistoryAction(
        db: Database,
        name: String,
        description: String,
        icon: String,
        liveData: SingleLiveData<Results>
    ) {
        try {
            val action = Action(
                invite = null,
                name = name,
                description = description,
                icon = icon,
                status = ActionStatus.HISTORIZED.toString()
            )

            db.actionDao().insertAll(action)
            liveData.postValue(Results.ACTION_SUCCESS)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            liveData.postValue(Results.ACTION_FAILURE)
        }
    }

    suspend fun addToHistory(
        actionId: Int,
        description: String,
        db: Database,
        liveData: SingleLiveData<Results>
    ) {
        try {
            val action = db.actionDao().getActionsById(actionId)
            action.status = ActionStatus.HISTORIZED.toString()
            action.description = description
            db.actionDao().update(action)
            liveData.postValue(Results.REJECT)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
            liveData.postValue(Results.FAILURE)
        }
    }
}